src/ObjCRuntime/Registrar.cs

```diff
@@ -556,9 +595,115 @@ public ObjCMethod (Registrar registrar, ObjCType declaringType, TMethod method)
 				}
 				set {
 					parameters = value;
+					native_parameters = null;
+				}
+			}
+
+			public TType [] NativeParameters {
+				get {
+					if (native_parameters == null && Parameters != null) {
+						// Put the parameters in a temporary variable, and only store them in the instance field once done,
+						// so that if an exception occurs, the same exception will be raised the next time too.
+						var native_parameters = new TType [parameters.Length];
+						for (int i = 0; i < parameters.Length; i++) {
+							var originalType = Registrar.GetBindAsAttribute (this, i)?.OriginalType;
+							if (originalType != null) {
+								if (!IsValidToManagedTypeConversion (originalType, parameters [i]))
+									throw Registrar.CreateException (4172, Method, $"The registrar can't convert from '{Registrar.GetTypeFullName (parameters [i])}' to '{originalType.FullName}' for the parameter '{Registrar.GetParameterName (Method, i)}' in the method {DescriptiveMethodName}.");
+								native_parameters [i] = originalType;
+							} else {
+								native_parameters [i] = parameters [i];
+							}
+						}
+						this.native_parameters = native_parameters;
+					}
+					return native_parameters;
+				}
+			}
+
+			bool IsValidToManagedTypeConversion (TType inputType, TType outputType)
+			{
+				var nullableType = Registrar.GetNullableType (outputType);
+				var isNullable = nullableType != null;
+				var arrayRank = 0;
+				var isArray = Registrar.IsArray (outputType, out arrayRank);
+
+				TType underlyingOutputType = outputType;
+				TType underlyingInputType = inputType;
+				if (isNullable) {
+					underlyingOutputType = nullableType;
+				} else if (isArray) {
+					if (arrayRank != 1)
+						return false;
+					if (!Registrar.IsArray (inputType))
+						return false;
+					underlyingOutputType = Registrar.GetElementType (outputType);
+					underlyingInputType = Registrar.GetElementType (inputType);
+				}
+				var outputTypeName = Registrar.GetTypeFullName (underlyingOutputType);
+
+				if (Registrar.Is (underlyingInputType, Foundation, "NSNumber")) {
+					switch (outputTypeName) {
+					case "System.Byte":
+					case "System.SByte":
+					case "System.Int16":
+					case "System.UInt16":
+					case "System.Int32":
+					case "System.UInt32":
+					case "System.Int64":
+					case "System.UInt64":
+					case "System.nint":
+					case "System.nuint":
+					case "System.Single":
+					case "System.Double":
+					case "System.nfloat":
+					case "System.Boolean":
+						return true;
+					default:
+						return Registrar.IsEnum (underlyingOutputType);
+					}
+				} else if (Registrar.Is (underlyingInputType, Foundation, "NSValue")) {
+#if MMP || MONOMAC
+					// Remove 'MonoMac.' namespace prefix to make switch smaller
```

Comment:
This is cute. I'll have to remember this trick (though hopefully classic isn't long for this world).